### ISSUE TITLE
"There is no disk in the drive" error even when not using CD's

### ISSUE THREAD

**Issue by InsufferableHexavaxxedNPC** [dosbox-staging/dosbox-staging#3026]
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### What Operating System are you using?

Windows 7

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

![nocderror](https://github.com/dosbox-staging/dosbox-staging/assets/129743500/b29a9e66-1af0-4bc3-8e8f-c6f7f634f697)
Greetings.

The problem is rather simple. I'm using a CD-ROM drive emulator called ImgDrive (https://www.yubsoft.com/imgdrive), and whenever I unmount a disc image, the drive letter which it used for mounting, is removed from registry, and if I so happen to run Staging after unmounting, it will throw the following error upon starting: "There is no disk in the drive. Please insert a disk into drive D:.", and though it goes away by clicking on either "Continue" or "Cancel" a few times, it tends to get annoying after a while. 

Now obviously this would be the correct, intended behaviour if I had set up a CD-ROM drive from Staging itself and it were using the host's drive as its own, but that's the thing; this happens even when I'm not utilizing Staging's CD-ROM functionality at all.

After rebooting, everything goes back to normal, until I have to mount/unmount something in ImageDrive again.

Best regards and sorry for any inconveniences.

### Steps to reproduce the behaviour.

Explain how to reproduce
1. 
2.
3.

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

**Comment by MasterO2** [comment]
> The problem is rather simple. I'm using a CD-ROM drive emulator called ImgDrive

https://github.com/dosbox-staging/dosbox-staging/wiki/Frequently-asked-questions#do-not-use-daemon-tools-to-mount-cd--disc-images-in-dosbox-staging

If you do use Daemon Tools or other similar programs (like ImgDrive) with Dosbox Staging, you **will** run into game compatibility issues [like this](https://github.com/dosbox-staging/dosbox-staging/issues/2465). In order to reduce the future likelihood of running into said compatibility issues, use `imgmount` only.

For further reference, please read https://github.com/dosbox-staging/dosbox-staging/wiki/DOS-commands#IMGMOUNT or while in Dosbox-Staging, type `imgmount /?` and press enter.

**Comment by InsufferableHexavaxxedNPC** [comment]
I don't use ImgDrive with Staging at all (except that one time when installing Carmageddon Splat Pack, which required CD swapping during installation), but I get to use it constantly for other things that I do on this computer. The problem is that Staging expects a CD-ROM drive to be there if it detected it earlier, which, again, wouldn't be an issue if I were actually mounting ImgDrive's virtual drive from Staging, which I'm not.

**Comment by johnnovak** [comment]
Super strange. Never heard of anything like this. I'm wondering if SDL is doing some shady things under the hood with drives? I have no idea what's going on, honestly... Does it happen with a clean config by running `dosbox -noprimaryconf -nolocalconf`. Just try to start up DOSBox like that from the console, do you still get the error?

**Comment by InsufferableHexavaxxedNPC** [comment]
Yep, happens with default config, fresh installation, reboot and all. Believe me I'm very careful about these things, particualrly about uninstalling and cleaning everything up before upgrading any piece of software, and I also assumed that leaving "Your configuration" blank would have been interpreted as "default, unmodified", but I guess it was a bit of a stretch. Sorry about that.

**Comment by kcgen** [comment]
@InsufferableHexavaxxedNPC , where did you download your DOSBox Staging package from?

Also, grepping our code for those error strings in that dialog box are coming up empty. (Perhaps the dialog is generated by Windows as a result of some CDROM API error?)

**Comment by johnnovak** [comment]
> Also, grepping our code for those error strings in that dialog box are coming up empty. (Perhaps the dialog is generated by Windows as a result of some CDROM API error?)

Yeah, hence I'm suspecting that backported SDL1 special thing to give us physical CD-ROM support on Windows.

**Comment by weirddan455** [comment]
I'll try to reproduce this later.  I think it is an issue with SDL CD.  Probably not an issue with ImgMount specifically.  I just noticed I get the following message in my `dmesg` kernel log in Linux every time I start up Dosbox (because I don't currently have a disc in my drive).

```
[42300.005521] sr 3:0:0:0: [sr0] CDROM not ready.  Make sure there is a disc in the drive.
```

There's some code in `SDL_CDROMInit` (gets called at Dosbox init regardless of if you've actually mounted any CDs) that scans all the drives on the computer looking for CD drives.  It's probably holding some Windows handle open and then gets upset with the disc gets ejected.

Like I said before, I would like to get rid of this library all together eventually (some times after the 0.81 release) but until then I'll see if I can find a way to patch it.  This is obviously not the behavior we want.

**Comment by weirddan455** [comment]
Oh I missed this part.  I was thinking you were simply unmounting while Staging is running.

>  whenever I unmount a disc image, the drive letter which it used for mounting, is removed from registry, and if I so happen to run Staging after unmounting, it will throw the following error upon starting

>  After rebooting, everything goes back to normal, until I have to mount/unmount something in ImageDrive again.

@InsufferableHexavaxxedNPC can you please confirm reproduction instructions?  As I now understand it:

1. Mount some .iso with ImgDrive (assume virtual drive is D:)
2. Unmount the .iso (this causes the D: drive to disappear from Windows Explorer?)
3. Start Dosbox Staging
4. Error appears stating there is no disk in the D: drive which is now hidden from Windows.

It's strange that it does that hiding.  IIRC DaemonTools leaves the drive in place and handles ejecting just like a physical drive.  As in, the drive is always visible to the OS.  It's just in an ejected state.  At least I think that's the way it's supposed to work.  It's been a while since I've used these tools.

**Comment by Grounded0** [comment]
> Like I said before, I would like to get rid of this library all together eventually (sometime after the 0.81 release) but until then I'll see if I can find a way to patch it. This is obviously not the behavior we want.

I remember this was removed with SDL2 transition and restored back to restore compatibility with games like Under a Killing Moon. We got our own disc image mounter and everything outside of it can be considered unsupported or we have very little interest in trying to resolve issues with them.

**Comment by Grounded0** [comment]
@InsufferableHexavaxxedNPC 

Multiple CDs can be mounted at a same time by specifying disc1.iso disc2.iso disc3.iso... and using Swap Image hotkey to change CDs which eliminates need for external image mounter in your case.

**Comment by weirddan455** [comment]
>  We got our own disc image mounter and everything outside of it can be considered unsupported or we have very little interest in trying to resolve issues with them.

@InsufferableHexavaxxedNPC is using this virtual CD drive for other purposes.  They're not actually mounting it inside Dosbox.  Dosbox is doing something weird by popping up this error.  This is our bug and I'm looking into it now.

>  I remember this was removed with SDL2 transition and brought back to restore compatibility with games like Under a Killing Moon.

I'm aware of the history.  I would prefer to replace SDL CD with direct WIN32 calls rather than remove support for physical CDs and virtual drives all together.  This would also fix CD audio not going through our mixer and maybe some of the MSCDEX bugs that have been mentioned.

For now, I'm just going to poke through the SDL CD init code and see if I can find a quick patch to fix this bug.

**Comment by weirddan455** [comment]
@InsufferableHexavaxxedNPC I am unable to reproduce the issue.  Please provide specific reproduction instructions.

I downloaded the ImgDrive software and tried as I said above.  I also noticed an option in Tools->Hide Empty Drives which does the hiding you described.  I tried with that both toggled on and off.  I even tried unmounting the ,iso inside ImgDrive while Dosbox has it mounted and I still do not see that pop-up.

I tested with both latest `main` branch and with 0.80.1.

Note that I am on Windows 10.  It's unfortunate if this is a Windows 7 specific bug as I don't have such a system to test on.

**Comment by InsufferableHexavaxxedNPC** [comment]
>"where did you download your DOSBox Staging package from?"

From the official Staging git. Back in the day, I'd occasionally hit EmuCR (recommended by the official DosBox wiki, but reckoned to be a bad idea for Staging particularly) because I have to be signed into GitHub in order to download the dev builds, and GitHub won't let me use its services anonymously so I get nagged to hell with account verification messages and that can be somewhat of a discouragement at times.

>"Can you not use that software? Why not just use DaemonTools lite or something."

That thing's bloated useless and filled to the brim with malware and God knows what else (always has). ImgDrive on the other hand, is lightweight, fast, stable, clean and works like a charm. Besides, it's also about the only other Windows CD-ROM emulator that can handle RedBook audio. I'm afraid the switch is not an option, but again, this is not a super horrible problem that's destroying the utility of Staging for me; it's barely a cosmetic annoyance.

>"can you please confirm reproduction instructions? As I now understand it:

>Mount some .iso with ImgDrive (assume virtual drive is D:)
>Unmount the .iso (this causes the D: drive to disappear from Windows Explorer?)
>Start Dosbox Staging
>Error appears stating there is no disk in the D: drive which is now hidden from Windows."

You must run Staging at least once before unmounting, then close Staging, then unmount and run Staging again to get the error here.

>"At least I think that's the way it's supposed to work."

I think an older OS like this wouldn't normally expect CD-ROM drives, hard drives and similar "permanent" devices to be plugged in and out on the fly like a usb drive or something, so ultimately a level of interoperability issues are bound to arise.

>"I would like to know behavior on your system when you try that"

I get the same error as the one on your screenshot.

**Comment by johnnovak** [comment]
> I'm afraid the switch is not an option, but again, this is not a super horrible problem that's destroying the utility of Staging for me; it's barely a cosmetic annoyance.

Yeah, so that's a personal choice, and if we don't find anything obvious, probably we'll just stop there and you'll need to live with this minor annoyance. 

We kinda/sorta support Windows 7, but as no one in the team is actively using it, it's not exactly first class support. I'm on Win 10, other are on Win 10 too or 11.

**Comment by Grounded0** [comment]
I use paid copy of DAEMON Tools for macOS and its not loaded up with crapware. I presume the same applies for the Windows version.

DAEMON Tools Lite can be bought from 3.99 EUR + taxes with lifetime updates at:

https://www.daemon-tools.cc/products/dtLite

People need money to eat or be landowners with arable land under capitalism so i've funded this project in tune of 13,000 USD to date. 🤷‍♂️

**Comment by Grounded0** [comment]
This is the premier Windows driver for disc image mounting used by DAEMON Tolls and Alcohol 120%/Alcohol 52%. Basically everything else is going to suck:

https://en.wikipedia.org/wiki/SCSI_Pass-Through_Direct

**Comment by kcgen** [comment]
At the time of testing the SDL CD patch, there were multiple people who verified it's behavior using physical CDROMs in Linux, macOS, and Win10 machines.

I personally burned a large stack of CDs.. for the following games, and testing them all versus `imgmount`. Here were my initial results, which Wengier dutifuly got working:

| Game  | main, no-usecd  | PR `-usecd 0`  | main + `imgmount cue` |
|---|:---:|:---:|:---:|
| Under a Killing Moon  | Install: 🗹 Play: ❌  |  Install: 🗹 Play: ❌  |  Install: 🗹 Play: 🗹  |
|  Ultima 7 The Complete Edition |  Install: 🗹 Play: 🗹  |  Install: 🗹 Play: 🗹  |  Install: 🗹 Play: 🗹  |
| The Pandora Directive  |  Install: 🗹 Play: ❌   |  Install: 🗹 Play: ❌  |  Install: 🗹 Play: 🗹  |
|  System Shock - Enhanced Edition | Install: 🗹 Play: ❌   |  Install: 🗹 Play: 🗹  |  Install: 🗹 Play: 🗹  |
| Space Quest - Collector's Edition  |  Install: 🗹 Play: 🗹  | Install: 🗹 Play: 🗹  | Install: 🗹 Play: 🗹  |
| Realms of the Haunting  |  Install: 🗹 Play: 🗹  | Install: 🗹 Play: 🗹  | Install: 🗹 Play: 🗹  |
| Gabriel Knight - Sins of the Fathers  |  Install: 🗹 Play: 🗹  | Install: 🗹 Play: 🗹  | Install: 🗹 Play: 🗹  |

You can see screenshots here: https://github.com/dosbox-staging/dosbox-staging/pull/1540#issuecomment-1034474763

There were many runs with and without CDROMs available, and never got this pop-up from Windows 10.

Are there code tweaks that could /possibly/ prevent windows 7 from creating this nussiance pop-up? Yes - probably.

But even to work on this, it would involve testing real physical CD access on a Windows 7 machine, which has been EOL for > 3.5 years.

On top of that, any changes would require re-testing the change on Windows 10.

This is a non-trivial (physical setup, testing, regression testing) all for an EOL system.  None of the maintainers even have a Windows 7 system.

@InsufferableHexavaxxedNPC, PRs /are/ welcome, so if this is something you want to do yourself - as well as all the regression testing on Windows 10, then it would be very welcome.

Additionally- if you want to commission a 3rd-party Windows 7/10 programmer do undertake this effort on your behalf, we have no qualms with who submits the PR.
